### DISCUSSION TITLE
Roadmapping 3.x

### DISCUSSION THREAD

**Discussion by jbrooksuk** [cachethq/cachet#4349] [General]
Hey everyone,

Learning from past mistakes, I want to work with the community to roadmap Cachet 3.x. It’s been a couple of weeks since I acquired Cachet back and a lot has been achieved already:

1. Launched a [new website](https://cachethq.io). We’ve done away with the Google Analytics & Facebook Pixel tracking — we’re now just keeping it simple with [Fathom](https://usefathom.com/james).
2. Launched [new branding](https://blog.cachethq.io/new-version-new-branding/). [Jump24](https://jump24.co.uk) were awesome and immediately came to help me with this. I think they’ve done a great job and I’m really pleased with the results.
3. [Open-sourced the new `cachethq/core` project](https://blog.cachethq.io/cachet-core/). This is where we’re going to focus our efforts (in terms of this roadmap). The `cachethq/cachet` project will install and pre-configure Cachet Core for you, making updates much easier for everyone involved.

## Considerations
Before we move into the roadmapping, I want to list some considerations that we need to keep in mind while developing Cachet.

1. The upgrade process should be as smooth as possible. We’re only going to be supporting PHP 8.1 and above, so of course that is a hurdle existing users need to get over, but with the right dependencies installed, we should be able to take an existing Cachet database and migrate to 3.x with as little effort as possible required.
2. Cachet should remain multi-lingual. That’s to say that all text should be put into a translation file. I’m personally leaning to only shipping `cachethq/core` with the `en-us` translations. This means we’re able to update translations separately, without needing to make a Cachet release each time.
3. Cachet should be tested. If a bug is fixed, a test should go with it. We’ll do our absolute best to prevent regressions.
4. 3.x should ship with additional features and bug fixes. I don’t think it’s enough to ship Cachet requiring an updated version of PHP, Laravel and a new frontend. We need some of those missing features and bug fixes.
5. Jump24 are also taking the time to design a new status page and dashboard for Cachet. We can start working on stubbing out, but essentially it’ll require the same information as 2.x — a status bar, component groups, metrics, incidents & updates etc.
6. The status page components should be publishable and customisable. Cachet will provide ways of changing colours (using CSS variables) but this would allow for more customisation that previously possible, maybe even the concept of themes? 👀 
7. Cachet should be extensible. We must provide events that can be hooked into by packages (`IncidentCreated`, `IncidentStatusChanged` etc).
8. Cachet needs to continue to work with SQLite, MySQL, Postgres and SQLServer. This isn't actually as easy as you'd hope...

## Roadmapping
I’ve listed out these points in a loose order and at a “high level”. Once we have _direction_, we can break each item down into smaller “projects” and hopefully work on some of them asynchronously.

1. Scaffold the `cachethq/core` project including; controllers, actions, migrations, models and actions. I’ve already started this to some extent and this should now provide enough of a starting point for people to contribute to.
2. Update the database. There are several columns in the Cachet database that don’t provide default values, don’t accept `null` and it’s created a confusing structure. We should make this more intuitive for us developers.
3. Stub out the status page. *I think we’ll use Blade and Livewire for this. We’ll need some interactivity such as tooltips and metrics, so we may want to use Alpine.js or something to go with it.*
4. Stub out the dashboard. *I think we’ll use Vue and Inertia for this. In my opinion, it’s the best stack for creating interactive dashboards.*
5. Stub out the API. I’ve already started this but we’ll need to create all of the routes, implement the HTTP API resources etc.
6. Work on the actions. Again, I’ve started this, but we’ll need to make sure we can create Incidents, Components, Component Groups etc.
7. Implement the designs from Jump24. These will be provided to us in a Figma file, so I’d love for this to be implemented with Tailwind.
9. Testing. Lots of people have offered to test Cachet 3.x already, so we’ll get this arranged as we get closer to releasing 3.0 betas.
10. Documentation. We also have a new documentation repo ([`cachethq/docs`](https://github.com/cachethq/docs)). We’ll use this and write fresh documentation for Cachet 3.x.
11. Launch! 🎉 

This is just a high-level overview of a roadmap. If we can agree on the points here, then we can flesh them out into individual projects and put them into GitHub projects. Perhaps with some community ownership of each part?

Let me know what you’re thinking! 🙂

**Comment by ricardov03** [comment]
Hello @jbrooksuk, I'm reading on Point 3 using Livewire and Point 4 with Vue and Inertia. Why use two different stacks? Is it to involve more community members or to take the opportunity to learn about both stacks?

**Reply by jbrooksuk** [reply]
👋🏻 Yeah, so I actually don't think that Inertia is a good fit for the status page itself since it'll rely on JS to render the page. Of course, we can use Inertia SSR but that's an additional step needed during set up.

**Comment by msh1596** [comment]
Hello @jbrooksuk , I think at high-level roadmapping is okay. I am ready to help on Livewire, Vue or Tailwind topics. So whenever the Projects table are updated, I can take a look on it.

**Reply by jbrooksuk** [reply]
Thanks! We'll get the Projects opened soon.

**Comment by SecondeJK** [comment]
I'm better at squashing bugs and investigative debugging work than architecting, which I can jump on pretty quickly.
But specifically, my thing is documentation and user experience. If I can get on that, write a strategy for it, or wireframe a docs part of it on GHpages then I'll do it if you're up for it.

**Comment by thebootable** [comment]
Sounds great! It's rare to be able to be involved in such an early stage.
Testing / testmanagement is something I'm good at as well as UX and docs. Though I'm not too familiar with the mentioned stack I'd love to contribute!
